### PR TITLE
MINOR: Separate GH workflows for PRs and trunk

### PR DESCRIPTION
**Pull request by mumrah**
Github has different behavior for PRs within apache/kafka and those from public forks. For security reasons, ASF Infra disallows forks to trigger workflows using the `pull_request` event (as this would expose repository secrets). Instead, we can use the `pull_request_target` event which triggers when a fork creates a PR. This event will cause the workflow to run using the code of the PR target (i.e., trunk).

Per the ASF Infra policy https://infra.apache.org/github-actions-policy.html, we must take care not to inject secrets into `pull_request_target` workflows. For this reason, I've created a separate "Public Pull Requests" workflow that does not public Develocity build scans.

For now the new fork workflow just does a compile check.

### CODE REVIEW

**Comment by chia7712** [comment]
> For security reasons, ASF Infra disallows forks to trigger workflows using the pull_request event (as this would expose repository secrets). Instead, we can use the pull_request_target event which triggers when a fork creates a PR. This event will cause the workflow to run using the code of the PR target (i.e., trunk).

not sure whether we need another ci file. we can change the limit of branches (https://github.com/apache/kafka/pull/16931#issuecomment-2298977317) to fix the issue (see https://github.com/apache/kafka/pull/16939 for non-committer's PR)

**Comment by chia7712** [comment]
> Per the ASF Infra policy https://infra.apache.org/github-actions-policy.html, we must take care not to inject secrets into pull_request_target workflows. For this reason, I've created a separate "Public Pull Requests" workflow that does not public Develocity build scans.

oh, that is a good point

**Comment by chia7712** [comment]
>  we must take care not to inject secrets into pull_request_target workflows.

not sure whether I misunderstand the "inject". IIRC, `secrets` is a safe way recommended by Github (https://docs.github.com/en/actions/security-for-github-actions/security-guides/using-secrets-in-github-actions), and other apache projects are using that to enable some extra services (gradle dvelocity, CodeCov, etc.)

in short, I don't think `secrets` will "export" the token but it does "inject" the secrets.

**Comment by mumrah** [comment]
@chia7712 I've updated this to have two workflows.

trunk
-----
* runs on each commit pushed to trunk
* runs compile check for 8, 11, 17, and 21
* runs the full test suite for 11 and 17
* publishes build scan to develocity
* writes to gradle cache
* lets previous action run finish

PR
---
* runs on each PR (opened, synchronized, reopened)
* runs compile check for 8, 11, 17, and 21
* runs the full test suite for 11 and 17
* publishes build scan to develocity only if PR from apache/kafka
* doesn't write to gradle cache (reads from it)
* only runs test suite if branch name is "gh-*"
* interrupts any action run for the current PR

I think this will let us start gaining some signal from the Actions builds without overwhelming the ASF runners just yet :) 

WDYT?

File: .github/workflows/pr.yml
Comment (chia7712) [inline .github/workflows/pr.yml:72]: Do we open PRs from apache/kafka? it seems almost all PRs are from public forks.

Also, the failed tests in PR phase are expected. Maybe publishing the scans in CI phase is more acceptable since that is what running in code base.

File: .github/workflows/ci.yml
Comment (chia7712) [inline .github/workflows/ci.yml:74]: Maybe we don't need retries in CI? That can helps us to find out more flaky

File: .github/workflows/pr.yml
Comment (chia7712) [inline .github/workflows/pr.yml]: in the https://github.com/apache/kafka/pull/16936#issuecomment-2299395186, you say PR will run compile check for all JDKs

**Review by chia7712 — COMMENTED** [review]
@mumrah thanks for this great PR! 

> only runs test suite if branch name is "gh-*"

Is this a flag which enables developers to select their favor CI? "gh-*" run on Github and others run on Jenkins?

File: .github/workflows/ci.yml
Comment (mumrah) [reply .github/workflows/ci.yml:74]: I agree, but this is how Jenkins is run today. Let's save this for a future PR :)

File: .github/workflows/pr.yml
Comment (mumrah) [reply .github/workflows/pr.yml:72]: I think all our PRs come from forks normally. Leaving this in here does give committers a way to get a build scan from a non-trunk code base.

**Comment by mumrah** [comment]
The purpose of the "gh-" filter is to give developers a way to opt-in to Github actions while we are evaluating it. It won't affect the Jenkins build.

We could also use a PR label, but only committers can set those.

**Comment by chia7712** [comment]
> The purpose of the "gh-" filter is to give developers a way to opt-in to Github actions while we are evaluating it. It won't affect the Jenkins build.

That makes sense. BTW, how to announce this preview feature to developers to get enough feedback/evaluation? Send email to dev channel or add it to `PULL_REQUEST_TEMPLATE.md`?

**Review by chia7712 — APPROVED** [review]

**Comment by mumrah** [comment]
> That makes sense. BTW, how to announce this preview feature to developers to get enough feedback/evaluation? Send email to dev channel or add it to PULL_REQUEST_TEMPLATE.md?

Both of those sound good. Once we get it working :)

File: .github/workflows/ci.yml
Comment (chia7712) [inline .github/workflows/ci.yml:47]: I notice that you open https://issues.apache.org/jira/browse/KAFKA-17381, so maybe we can remove `--info` to reduce the log output?

**Comment by mumrah** [comment]
@chia7712 FYI after merging this, I saw that https://github.com/apache/kafka/pull/16943 needed to manually approved before it could run.

**Pull request merged**
